### PR TITLE
prov/tcp: adding a port_range variables

### PR DESCRIPTION
**Pull request by nikitaxgusev**
Added options `FI_TCP_PORT_HIGH_RANGE`  and `FI_TCP_PORT_LOW_RANGE` in tcp provider in order to use a specific range of ports for launching applications.
 
Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>

### CODE REVIEW

**Comment by shefty** [comment]
Can you please add a description to the commit message why we would want this change?

**Comment by nikitaxgusev** [comment]
@shefty Commit message is ready.

**Comment by shefty** [comment]
I understand *what* the patch is doing.  What I don't understand is *why* we want this.

**Comment by nikitaxgusev** [comment]
@shefty  There is a user request to provide ability to limit range of ports used by TCP provider. That covers the situation when only some limited range of ports is available to application _(e.g.: firewalls)_.
 So, we can handle such situations using `FI_TCP_PORT_RANGE` environment variable, which give us a possibility to define the used range.

**Comment by vkrishna** [comment]
@nikitaxgusev you can pass a port number in hints for getinfo, if you want to use a particular port number. If fi_getinfo fails for that port you can cycle through ports in port range.

**Comment by shefty** [comment]
It seems that this sort of limitation really needs to come from the OS itself, rather than trying to rely on apps to self limit their dynamic port selection.

**Comment by ddurnov** [comment]
@shefty I beleive that the firewall settings are non local in most cases. It is on some router in the network. Local node knows nothing about availalbe set of ports in that case.
@vkrishna It looks like something very provider specific, right? We try to avoid any provider specific code on our side (Intel MPI). In case it is somethnig generic for all providers, then we may have a generic knob for the port range (E.g. may be required by some providers for connection establishment logic)

File: prov/tcp/src/tcpx_ep.c [hunk 1]
```diff
@@ -747,3 +808,56 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
...
+	int  high;
+	int  low;
+	int  delim;
+
+	if (fi_param_get_str(&tcpx_prov, "port_range", &range)) {
+		return -FI_ENODATA;
+	}
+
+	char *str = range;
+	for (delim = 0; range[delim]; range[delim]==':' ? delim++ : *range++);
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Move stand alone semi-colons to their own line.

**Comment by nikitaxgusev** [comment]
@shefty  Please, can you recheck the updated pull-request?
I have the fail with Intel CI, can you say what's wrong there, I haven't got the access?
As for the pull-request, I've updated the description of commit and provided requested changes.
Thanks.

**Comment by shefty** [comment]
The Intel CI looks unrelated.  @a-ilango - you may want to look at it.  An assertion is being hit.

File: prov/tcp/src/tcpx.h [hunk 2]
```diff
@@ -120,6 +127,7 @@ struct tcpx_pep {
 	struct fi_info		*info;
 	SOCKET			sock;
 	struct tcpx_cm_context	cm_ctx;
+	struct tcpx_port_range	port_range;
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: We don't need to store this per passive endpoint.  A single global for the module is sufficient.

File: prov/tcp/src/tcpx_ep.c [hunk 3]
```diff
@@ -65,6 +66,20 @@ void tcpx_hdr_bswap(struct tcpx_base_hdr *hdr)
...
+
+	ret = setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, (char *) &optval,
+			 sizeof(optval));
+	if (ret) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt nodelay failed\n");
+		return ret;
+	}
+
+	return ret;
+}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This change is unrelated to restricting the port value.  Either drop or move into a separate commit that explains the change.

File: prov/tcp/src/tcpx_ep.c [hunk 4]
```diff
@@ -185,6 +200,42 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
...
+					strerror(ofi_sockerr()));
+				goto err;
+			}
+		} else {
+			break;
+		}
+	}
+	return FI_SUCCESS;
+err:
+	ofi_close_socket(pep->sock);
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This function doesn't open the socket, so we shouldn't close it on failure.  Leave that to the caller.

File: prov/tcp/src/tcpx_ep.c [hunk 5]
```diff
@@ -185,6 +200,42 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
 	return ret;
 }
 
+static int tcpx_pep_bind_to_port_range(struct tcpx_pep *pep)
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This function can be generalized.  Instead of passing in the passive EP, pass in the socket and src address.  This will make it usable with active EPs, and make it easier to extract into common code if other providers need the same functionality.

File: prov/tcp/src/tcpx_ep.c [hunk 6]
```diff
@@ -185,6 +200,42 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
...
+{
+	int ret, i, rand_port_number;
+
+	srand(getpid());
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: sand() requires including stdlib.h.  Please add that as an include file.
We don't want to call srand() on every call into this function.  That will reset the random number generator to the same starting value each time.  Move srand() into the tcp initialization function.

File: prov/tcp/src/tcpx_ep.c [hunk 7]
```diff
@@ -185,6 +200,42 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
...
+		ret = bind(pep->sock, pep->info->src_addr,
+		      (socklen_t) pep->info->src_addrlen);
+		if (ret) {
+			if (errno == EADDRINUSE) {
+				continue;
+			} else {
+				FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+					"failed to bind listener: %s\n",
+					strerror(ofi_sockerr()));
+				goto err;
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Use
```
return -errno
```
 to be consistent with other functions, where negative errno is returned in case of failures.

File: prov/tcp/src/tcpx_ep.c [hunk 8]
```diff
@@ -185,6 +200,42 @@ static int tcpx_ep_shutdown(struct fid_ep *ep, uint64_t flags)
...
+			}
+		} else {
+			break;
+		}
+	}
+	return FI_SUCCESS;
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: This needs to be something like
```
return (i < port_range.high) ? FI_SUCCESS : -FI_EADDRNOTAVAIL;
```

File: prov/tcp/src/tcpx_ep.c [hunk 9]
```diff
@@ -209,18 +260,39 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
...
+			ret = tcpx_setup_socket(pep->sock);
+			if (ret) {
+				goto err;
+			}
+		} else {
+			ret = tcpx_setup_socket_nodelay(pep->sock);
+			if (ret) {
+				goto err;
+			}
+		}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Why isn't tcpx_setup_socket() just called in all cases?

File: prov/tcp/src/tcpx_ep.c [hunk 10]
```diff
@@ -209,18 +260,39 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
...
+		ret = bind(pep->sock, pep->info->src_addr,
+		      (socklen_t) pep->info->src_addrlen);
+
+		if (ret) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: %s\n",
+				strerror(ofi_sockerr()));
+			goto err;
+		}
+	} else {
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Please change this to
```
else if (port_range.high) {
    tcpx_bind_to_port()...
}
```
If a port range hasn't been specified, let the OS continue to pick it and avoid the overhead of the provider trying to find one.

File: prov/tcp/src/tcpx_ep.c [hunk 11]
```diff
@@ -747,3 +825,46 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
...
+	}
+
+	if (high < low) {
+		if (high == 0 &&  high < low) {
+			high = TCPX_PORT_MAX_RANGE;
+		} else {
+			low  = 0;
+			high = 0;
+		}
+	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Discard all invalid input
```
if (low < 0 || high < 0 || low > high) {
    FI_WARN(....);
    return;
}
```
I'd be fine with this, however:
```
if (high > TCP_PORT_MAX)
    high = TCP_PORT_MAX;
```
This way, the user doesn't need to define low at all, or know what the true maximum port value is.  The resulting range of values will still be correct.

File: prov/tcp/src/tcpx_ep.c [hunk 12]
```diff
@@ -731,6 +803,12 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 	_pep->cm_ctx.cm_data_sz = 0;
 	_pep->sock = INVALID_SOCKET;
 
+	ret = tcpx_set_port_range(_pep);
+	if (ret == -FI_EINVAL) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"Invalid info\n");
+		return -FI_EINVAL;
+	}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Read the port range values where other provider environment variables are read, likely in init somewhere.

File: prov/tcp/src/tcpx_ep.c (hunk 9 above)
Comment (nikitaxgusev) [reply prov/tcp/src/tcpx_ep.c]: @shefty Since we select the port automatically at the provider level, we have to go through the range of ports and find the first free one.
In the case of `SO_REUSEADDR`, `bind ()` allows the port to be reused, which breaks the logic of port selection and as the results, it's a deadlock.

**Comment by nikitaxgusev** [comment]
@shefty 
Please, can you take a look of my last changes? What I did:

1. moved  the getting environment variables and `srand()` into tcpx_init.c

File: prov/tcp/src/tcpx_ep.c [hunk 13]
```diff
@@ -209,18 +254,31 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
...
+		}
+		ret = bind(pep->sock, pep->info->src_addr,
+		      (socklen_t) pep->info->src_addrlen);
+
+		if (ret) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to bind listener: %s\n",
+				strerror(ofi_sockerr()));
+			goto err;
+		}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: nit: Move this if check outside of the main if-else.

File: prov/tcp/src/tcpx_ep.c [hunk 14]
```diff
@@ -209,18 +254,31 @@ static int tcpx_pep_sock_create(struct tcpx_pep *pep)
...
-	if (ret) {
-		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
-			"failed to bind listener: %s\n",
-			strerror(ofi_sockerr()));
-		goto err;
+		ret = tcpx_bind_to_port_range(pep->sock, pep->info->src_addr,
+					      pep->info->src_addrlen);
+		if (ret){
+			goto err;
+		}
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Drop this if statement and use the one above moved down, so we get an error message in either case.

File: prov/tcp/src/tcpx_ep.c (hunk 3 above)
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Call this function from the other place(s) where NODELAY is set.

File: prov/tcp/src/tcpx_ep.c (hunk 14 above)
Comment (nikitaxgusev) [reply prov/tcp/src/tcpx_ep.c]: Done.

File: prov/tcp/src/tcpx_ep.c (hunk 13 above)
Comment (nikitaxgusev) [reply prov/tcp/src/tcpx_ep.c]: Done.

File: prov/tcp/src/tcpx_ep.c [hunk 15]
```diff
@@ -747,3 +811,29 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
...
+	if (high > TCPX_PORT_MAX_RANGE) {
+		high = TCPX_PORT_MAX_RANGE;
+	}
+
+	if (low < 0 || high < 0 || low > high) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"Invalid info\n");
+		return -FI_EINVAL;
+	}
+
+	if (high != 0 && low == 0) {
```
[the remark this answers is not in the document]
Comment (nikitaxgusev) [reply prov/tcp/src/tcpx_ep.c]: Allowed now.

File: prov/tcp/src/tcpx_ep.c (hunk 3 above)
Comment (nikitaxgusev) [reply prov/tcp/src/tcpx_ep.c]: @shefty  I've researched that we also call `tcpx_setup_socket()` in `tcpx_endpoint()`, where we can use `REUSEADDR` + `TCP_NODELAY` freely, and it seems to me, there is not places  to use clear  `nodelay`.

**Comment by shefty** [comment]
@vkrishna - please review and merge if this looks good to you.

File: prov/tcp/src/tcpx_ep.c [hunk 16]
```diff
@@ -747,3 +808,24 @@ int tcpx_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 	free(_pep);
 	return ret;
 }
+
+int tcpx_set_port_range ()
```
Comment (vkrishna) [inline prov/tcp/src/tcpx_ep.c]: Why is this called after selecting the port? Can we combine this function with tcpx_init_env() and call from init?
